### DISCUSSION TITLE
MergerFS mount point crashing

### DISCUSSION THREAD

**Discussion by tigerkzr** [trapexit/mergerfs#1158] [Q&A]
I was wonder if anyone has a guide or suggestions to get me started in the right direction towards trouble shooting why my mergerfs mount point keeps crashing.  

My setup is ubuntu 22.10 server on Dell server hardware and up until a few weeks ago, MergerFS was rock solid. Now I am experiencing a crash of the mount point every other day or so. Basically I get the "transport endpoint is not connected" error. I was originally using a self compiled binary from the latest source.  Since the problems started, I have reverted to the binary available from the Debian install package with no improvement.

At this point, I am just not sure where to even begin digging into it.  I have used grep to try to dig into the system and kernel logs for mergerfs and fuse but end up with no results.

Any help is appreciated

**Comment by trapexit** [comment]
If it is crashing then you won't really find anything in logs. If it was a known situation and logable it wouldn't be crashing :)

Could you provide basic details as requested here: https://github.com/trapexit/mergerfs/tree/latest-release#information-to-include-in-bug-reports

The only way to see exactly what is going on is to catch it when it fails but that would require attaching a debugger.

**Reply by tigerkzr** [reply]
Here's the rundown:

MergerFS Version:  `2.34.1`

fstab entry:
```
## MergerFS DataPool Disks
UUID=<uuid>     /blackbeard/datapool/D20T-<drv serial>      ext4 defaults 0 2
UUID=<uuid>     /blackbeard/datapool/D20T-<drv serial>      ext4 defaults 0 2
UUID=<uuid>     /blackbeard/datapool/D18T-<drv serial>      ext4 defaults 0 2
UUID=<uuid>     /blackbeard/datapool/D18T-<drv serial>      ext4 defaults 0 2
UUID=<uuid>     /blackbeard/datapool/D16T-<drv serial>      ext4 defaults 0 2
UUID=<uuid>     /blackbeard/datapool/D16T-<drv serial>      ext4 defaults 0 2
UUID=<uuid>     /blackbeard/datapool/D16T-<drv serial>      ext4 defaults 0 2

##MergerFS
/blackbeard/datapool/D* /blackbeard/data    fuse.mergerfs defaults,allow_other,use_ino,noforget,inodecalc=path-hash,fsname=data_pool,category.create=pfrd,cache.files=auto-full,cache.writeback=true,cache.symlinks=true,cache.readdir=true,cache.entry=120,cache.attr=120,cache.statfs=10,dropcacheonclose=true,moveonenospc=true      0 2

## NFS bind mount for /data share
/blackbeard/data /srv/export/data                        none bind     0 0

```
uname -a:
```
Linux <hostname> 5.19.0-38-generic #39-Ubuntu SMP PREEMPT_DYNAMIC Fri Mar 17 17:33:16 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

hostnamectl:
```
       Icon name: computer-server
         Chassis: server 🖳
      Deployment: development
        Location: Closet Under the Stairs
      Machine ID: <machine ID>
         Boot ID: <boot ID>
Operating System: Ubuntu 22.10
          Kernel: Linux 5.19.0-38-generic
    Architecture: x86-64
 Hardware Vendor: Dell Inc.
  Hardware Model: PowerEdge T430
Firmware Version: 2.16.0
```

df -h (all ext4):
```
Filesystem                        Size  Used Avail Use% Mounted on
/dev/sdf1                          15T   12T  2.8T  81% /blackbeard/datapool/D16T-<drv serial>
/dev/sdd1                          17T   14T  2.6T  85% /blackbeard/datapool/D18T-<drv serial>
/dev/sde1                          15T   13T  2.3T  85% /blackbeard/datapool/D16T-<drv serial>
/dev/sdc1                          17T   14T  3.4T  80% /blackbeard/datapool/D18T-<drv serial>
/dev/sdh1                          15T   12T  3.3T  78% /blackbeard/datapool/D16T-<drv serial>
/dev/sdb1                          19T   14T  3.4T  81% /blackbeard/datapool/D20T-<drv serial>
/dev/sda1                          19T   14T  3.5T  80% /blackbeard/datapool/D20T-<drv serial>
```

perm/owner/group for all mount points:
`00755 nobody:nogroup`

perm/owner/group for all 1st level dirs:
`02755  root:root`

perm/owner/group for 2nd and lower level dirs
`02755 1000:1000`

client apps are all `1000:1000` (docker containers user:group) and only access 2nd level dirs or lower. However, issue results in complete inaccessibility of mount to the system `transport endpoint is not connected` even on `cd` command.

unknown how to repeat, however, anecdotally, it occurs with increased disk activity such as during a snapraid sync.  I THINK I can trigger it by running a directory scan operation on the mergerfs mount point with a docker containerized version of qdirstat (jlesage/qdirstat:latest) but I don't have an ability to determine exactly when it occurs during a scan.  Sometimes it has finished a scan without issue, sometimes it crashes mid scan. However, the crashes started occurring before ever using qirdirstat and have continued with the qdirstat container removed from the system.  I have been unable to catch the crash as it happened.


again, any advice or direction is greatly appreciated.  If any additional information may help track this down, I will happily provide.

**Reply by trapexit** [reply]
There isn't much I can do but suggest a debug build and generating a core dump or catching it when it crashes using a debugger.

**Reply by trapexit** [reply]
Do you know of anything that may have changed in the setup? New kernel? New hardware setup? Does it only happen with NFS?

**Reply by skilleter** [reply]
I hit a very similar problem recently; the mergerfs mount point would become inaccessible on Ubuntu Server 22.04.2. 

I (hopefully) fixed it by uninstalling mergerfs, cloning the source tree from Git, checking out version 2.34.1 and building and installing it.

Before I did this, mergerfs was breaking on a daily basis for a week. Since I did it, about 2 weeks ago, I've had no issues with mergerfs.

I don't know what triggered the problem, as I don't recall making any configuration changes, so I presume it was an Ubuntu update was the (presumably indirect) cause of the issue.

John

**Reply by trapexit** [reply]
@skilleter How had you installed it prior?

**Reply by skilleter** [reply]
Pretty sure it was just via `apt install mergerfs`

**Reply by trapexit** [reply]
Hmm. That would be coming from Debian folks directly and is pretty much always outdated so it is difficult to comment without more details. There was an odd situation with NFS where it somehow resulted in a request for the parent of the root which wasn't handled in libfuse and caused a crash but I fixed that so that shouldn't be the problem with @tigerkzr.

**Reply by tigerkzr** [reply]
So while researching, I realized I have been oblivious(just wasn't paying attention) to what the current linux kernels were doing.  I'm currently on 5.19 which is now end of life.  I'm not 100% certain, but I think these issues started after a recent ubuntu kernel update on 5.19.  I will be updating to 6.2 and will report back if I can trigger the issue again.

**Reply by trapexit** [reply]
I just recently updated my primary data server from 20.04 (and 5.15) to 22.04 (and 5.19) and have had no issues though I have mostly been running a yet to be released version. Then again I don't use NFS which could be related.

I was about to merge one more change and then release 2.35.0. Can add that to your testing.

**Reply by tigerkzr** [reply]
Ooooh! New and shiny! I love new and shiny. 

I can not be even remotely sure that the 5.19 that Ubuntu patch a few weeks ago caused it, just that it approximately coincided. It was a standard apt update > apt upgrade that pulled it down. I just remember having to reboot for the kernel update and not long after the crashes started. Until that point, I had had absolutely zero issues with the mergerfs mount since I built this system in october, including the mergerfs mount to an NFS export. 

If I can make it happen again after moving to 6.2 I'll compile a debug version and get you what I can. Otherwise, if I am unable to recreate, I'll drop a note here to that effect.

**Reply by tigerkzr** [reply]
Kernel 6.2.8, mergerfs 2.35.0-compiled:

I believe I may have narrowed it down.  Everything is stable for hours until I start my SABnzbd container with download tasks in the queue. Within about 10 minutes of starting SAB, the mount point crashes. I have no idea what it's doing to cause it, but I'll take what little bit of progress I can get. At the moment, it appears that I can run nzbget without issue.

**Reply by trapexit** [reply]
I suggest compiling `make DEBUG=1` and then attach gdb to it and then trigger the crash.

`gdb attach <pid>`

when it crashes `thread apply all bt`

**Reply by tigerkzr** [reply]
I guess it would also help to mention that I am running linuxserver/sabnzbd:nightly

**Reply by tigerkzr** [reply]
will do on the debug.  I'll get back to you sometime tomorrow.

**Reply by tigerkzr** [reply]
ok, small issue. When I attach to the mergerfs process it kills any container that has a bind mount to the mergerfs pool. It doesn't crash the container or anything, it just kind of "stops".  The webUI loses connection and starting any container that mounts a volume to the pool hangs on "starting".

**Reply by trapexit** [reply]
Yes, attaching stops the execution of the app. You have to run "continue" or just "c" in the interface to get it to continue.

**Reply by tigerkzr** [reply]
ahhh.. lol ima noob

**Reply by tigerkzr** [reply]
here's what gdb produced.  

[gdb output.txt](https://github.com/trapexit/mergerfs/files/11117919/gdb.output.txt)

**Reply by trapexit** [reply]
How many CPUs does your system have? That's a lot of threads. Not that it should matter.

This was built with `make DEBUG=1`? It's missing details which I'd expect to see.

**Reply by tigerkzr** [reply]
48 logical cores :). Dual processor dell t430.

I used  "make DEBUG=1 deb"

**Comment by trapexit** [comment]
Oh, no. Just make DEBUG=1 and run build/mergerfs. Building the deb does its own thing.

**Reply by tigerkzr** [reply]
redoing now.  

PS thanks for your patience

**Reply by tigerkzr** [reply]
Don't really want to jinx it, but this might be the deb package.  I'm throwing everything at it that would typically crash it within minutes and it's going strong with the non-deb build. I'll let it continue overnight and if it hasn't crashed, I'll try again.

Every crash up till now has been on either a "make deb" build or the .deb package downloaded from releases.

**Reply by trapexit** [reply]
Curious. Most people, including myself, always use my deb packages or those from the distro repo. They do use a different set of compile options but shouldn't be anything that leads to crashes. I'll have to check what they do or maybe build a debug version of the deb.

**Reply by tigerkzr** [reply]
caught it. Just took a while for it to hit.

[gdb.output.txt](https://github.com/trapexit/mergerfs/files/11118841/gdb.output.txt)

**Reply by trapexit** [reply]
Have you changed your settings? It looks like you are using read and worker threads now.

**Reply by tigerkzr** [reply]
nope, same settings as I posted above

**Reply by trapexit** [reply]
OK, found a bug in the default settings but that shouldn't matter in terms of this issue. Inadvertently changed the default value for read-thread-count to -1 instead of 0.  

Still looking at the exact crash condition.

**Reply by tigerkzr** [reply]
if there's anymore you need dug into, please let me know. I appear to still be stable as long as I don't use the SABnzbd container.

**Reply by trapexit** [reply]
The place where it says it is crashing doesn't really make sense to me at the moment. Could you try setting `read-thread-count=0`  and testing again.

**Reply by tigerkzr** [reply]
yup...  I'll post back as soon as I get the crash again

**Reply by tigerkzr** [reply]
Relatively quick this time ~15mins

[gdb.output_read-thread-count_0.txt](https://github.com/trapexit/mergerfs/files/11125275/gdb.output_read-thread-count_0.txt)

**Reply by tigerkzr** [reply]
here's my fstab line for the mount for that last one
```
/blackbeard/datapool/D*     /blackbeard/data           fuse.mergerfs defaults,use_ino,allow_other,noforget,async_read=false,read-thread-count=0,inodecalc=path-hash,fsname=data_pool,category.create=pfrd,cache.files=auto-full,cache.writeback=true,cache.symlinks=true,cache.readdir=true,cache.entry=120,cache.attr=120,cache.statfs=10,dropcacheonclose=true,moveonenospc=true 0 2
```

**Reply by trapexit** [reply]
Is access direct or via NFS?

**Reply by tigerkzr** [reply]
direct.  the docker containers have volumes bind mounted to directories on the mergerfs mount.

**Reply by tigerkzr** [reply]
well.. just had a crash with SAB not running.  I had NFS going this time though.

**Reply by trapexit** [reply]
I've got my filesystem stress tool going against a tmpfs mount for hours now and nothing. Going to remove all caching to try to stress it more. According to the stack trace the kernel is telling mergerfs to use some data mergerfs stored prior and that memory is not valid. Unfortunately, without being able to replicate the problem it's gonna be a pain to track down.

**Reply by tigerkzr** [reply]
No worries. I'm mostly functional.  I turned off NFS just in case and I have alternates for the troublesome apps. Since it's my system doing this, please let me know if there's something I can do to help.

**Reply by tigerkzr** [reply]
So this makes me think it might be the Kernel after all.  I know the problems started just after a kernel update from ubuntu. Perhaps there's something in the change logs.  These are the three previous kernels I had on my system. I'm just not sure if it was the last one that did it or the middle one.

```
5.19.0.38.39                     Installed
5.19.0.38.34                     Installed
5.19.0.35.36                     Installed
```

I'm on 6.2.8 right now so possibly it got broken then propogated to the latest

**Reply by trapexit** [reply]
It is possible but I'm running Ubuntu 22.04 with 5.19.0-35-generic and I would guess many others are too.

**Reply by tigerkzr** [reply]
yah, it didn't break until one of the .38s

**Reply by trapexit** [reply]
OK. I have it installed but I guess haven't rebooted in a while. Once I get some time I can try booting up the 5.19.0-38.

**Reply by trapexit** [reply]
24+ hours of stressing and nothing. I'm busy getting ready for a major move but if I find time I'll try setting up a VM or a secondary machine to test with a newer kernel. Unfortunately, it could a very tight race condition or something in which case it could be I never hit it on my setup.

**Reply by tigerkzr** [reply]
Oh man, worry about your move.  As I said, things are mostly stable as long I avoid certain apps so my setup is far from crippled by this. No crashes since the last one I reported.  I definitely understand the complexities, and likely luck, in tracking down what's actually happening.   Further, as I seem to be the only one experiencing this so far, I can't exactly say it's a major priority for your resources. Given that, I greatly appreciate all the attention you have given this so far.

So again, worry about your move, and if you find the time, awesome, if you're too busy with other things, no big deal.  If you do find the time, please keep me updated and absolutely let me know if I can do anything to assist.

**Reply by tigerkzr** [reply]
Just an update: I haven't had a crash since the last one I posted about here. I have been avoiding using SAB, so I may give it a few more days to finish up some tasks then try to crash it with SAB again.
